lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+// lib/Tranforms/IPO/StructFieldCacheAnalysis.cpp - Performs Cache-Aware
+// Structure Analysis-*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===------------------------------------------------------------------------===//
+//
+// This files implements FieldReferenceGraph and CloseProximityGraph
+//
+// FieldReferenceGraph implements Field Reference Graph (FRG) that represents
+// relationships between two consecutive field accesses. It looks like Control
+// Flow Graph (CFG) except that each node represents a field access and each
+// edge between nodes represents how frequent the pair happens consecutively and
+// how many bytes of other memory accesses are between the pair on average.
+//
+// CloseProximityBuilder builds Close Proximity (CP) relations of every pair of
+// struct fields of this struct type. The CP relations can be obtained by adding
+// up all paths between two nodes in FRG with brutal force, which results in
+// exponential complexity. Instead, we use an FRG collapsing algorithm to
+// collapse FRG and update CP relations in the same time. The collapsing
+// algorithm is trying to summarize the CP relations between a nodes with all
+// nodes in its subtree into a table stored with node. With collapsing, a tree
+// can be reduced to a node with a table, as long as the CP relations between
+// nodes in the table is updated. Thus a FRG will eventually be reduced to a
+// root nodes with a relation table, or a much smaller tree with a few
+// non-collapsed nodes. In either case, the complexity is much smaller than
+// exponential that can complete with reasonable amount of time.
+//
+//===------------------------------------------------------------------------===//
+
+#include "StructFieldCacheAnalysisImpl.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/CFG.h"
+#include "llvm/Support/Format.h"
+
+#include <cmath>
+#include <queue>
+
+using namespace llvm;
+
+#define DEBUG_PRINT_COUNT(x) (format("%.2f", (x)))
+#define DEBUG_PRINT_DIST(x) (format("%.3f", (x)))
+
+static cl::opt<bool> PerformFRGOnly(
+    "struct-analysis-FRG-only", cl::init(false), cl::Hidden,
+    cl::desc("Stop the analysis after performing FRG generation"));
+
+static cl::opt<bool> PerformCPGCheck(
+    "struct-analysis-check-CPG", cl::init(false), cl::Hidden,
+    cl::desc("Perform CPG checking algorithm that takes a long time"));
+
+static cl::opt<bool> DisableIgnoreZeroCountNodes(
+    "struct-analysis-disable-ignore-zeros", cl::init(false), cl::Hidden,
+    cl::desc("Ignore zero-count nodes in FRG generation"));
+
+// Utility functions
+// Function used to update (count, distance) pair by taking a weighted average
+// of the Result and Source and save result to Result. Used when updating
+// CPG or updating an edge weight
+static void updatePairByMerging(ExecutionCountType &ResultCount,
```

Comment:
I can group them into pairs if needed. But it doesn't look like a trivial change to me either...